app/lib/import_export.rb

```diff
@@ -11,497 +12,186 @@ module ImportExport # rubocop:todo Metrics/ModuleLength
-              rct_node(:patients, 'Monitoring Period', %i[last_date_of_exposure continuous_exposure symptom_onset symptom_onset_defined_by
-                                                          extended_isolation end_of_monitoring closed_at monitoring_reason expected_purge_ts]),
-              rct_node(:patients, 'Reporting Info', %i[responder_id head_of_household pause_notifications last_assessment_reminder_sent])
-            ]
-          }
-        ]
-      }
-    ]
-  }.freeze
-
-  ASSESSMENTS_EXPORT_OPTIONS = {
-    label: 'Reports',
-    nodes: [rct_node(:assessments, 'Reports', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id symptomatic
-                                                 who_reported created_at updated_at symptoms])]
-  }.freeze
-
-  LABORATORIES_EXPORT_OPTIONS = {
-    label: 'Lab Results',
-    nodes: [rct_node(:laboratories, 'Lab Results', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id lab_type
-                                                      specimen_collection report result created_at updated_at])]
-  }.freeze
-
-  CLOSE_CONTACTS_EXPORT_OPTIONS = {
-    label: 'Close Contacts',
-    nodes: [rct_node(:close_contacts, 'Close Contacts', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id first_name
-                                                           last_name primary_telephone email contact_attempts last_date_of_exposure assigned_user notes
-                                                           enrolled_id created_at updated_at])]
-  }.freeze
-
-  TRANSFERS_EXPORT_OPTIONS = {
-    label: 'Transfers',
-    nodes: [rct_node(:transfers, 'Transfers', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id who from_jurisdiction
-                                                 to_jurisdiction created_at updated_at])]
-  }.freeze
-
-  HISTORIES_EXPORT_OPTIONS = {
-    label: 'History',
-    nodes: [rct_node(:histories, 'History', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id created_by history_type
-                                               comment created_at updated_at])]
-  }.freeze
-
-  CUSTOM_EXPORT_OPTIONS = {
-    patients: PATIENTS_EXPORT_OPTIONS,
-    assessments: ASSESSMENTS_EXPORT_OPTIONS,
-    laboratories: LABORATORIES_EXPORT_OPTIONS,
-    close_contacts: CLOSE_CONTACTS_EXPORT_OPTIONS,
-    transfers: TRANSFERS_EXPORT_OPTIONS,
-    histories: HISTORIES_EXPORT_OPTIONS
-  }.freeze
+  # Finds the symptoms needed for the reports columns
+  def update_assessment_symptom_data(data)
+    # Don't update if assessment symptom data isn't needed
+    return unless data.dig(:assessments, :checked)&.include?(:symptoms)
+
+    # Using Symptom.all for finding distinct symptom names is actually faster here because it can use symptoms_index_chain_1
+    symptom_names = Symptom.all.distinct.pluck(:name).sort
```

Comment:
I acknowledge that this changes the functionality of what the exports currently do but I'm wondering if it would be ok for us to include these other symptoms? Querying for distinct symptom name based on joining patients -> assessments -> conditions -> symptoms currently takes ~30 seconds locally for me with 36k monitorees, 85k assessments, and 2.3 million symptoms while it only takes 22ms for the unscoped Symptom query to run.

Reply:
Is this something we can analyze after the entire spreadsheet is put together to see if every row has an empty value and then delete it?

Reply:
I thought about that too but `fast_excel` doesn't do reads, so I'm not sure if we can do anything retrospectively without using another library to load it, which I think would cause a decent increase in memory

Reply:
> So I did another deep dive into this and basically what I've found is that querying for different symptom types separately saves memory (because we don't have to pluck type and only have to pluck that data type's value). However, this does slow things down due to there being multiple queries, all of which are already complicated in that they involve multiple joins, so I think it's probably best to query all of them at the same time: ab04729#diff-c34521fcd0fdc2034d456886aae8fd35299fe908ab88d8e49c60d0b4d39ab26bR379

> I left a comment in the code about my findings and this decision (happy to discuss though). I also tried an intermediate solution of plucking all the ids of those symptoms first (to hopefully avoid executing the nasty joins from the query multiple times) and then firing separate queries for each data type (to hopefully still get some memory improvements from it), but that turned out to be even slower than both existing solutions.

https://github.com/SaraAlert/SaraAlert/pull/717#discussion_r586003589